File: test/examples/basic_example/basic_example.cc































#include <cstdlib>
#include <memory>

#include "absl/flags/flag.h"
#include "absl/log/flags.h"
#include "ortools/base/init_google.h"
#include "ortools/base/logging.h"
#include "ortools/init/init.h"
#include "ortools/linear_solver/linear_solver.h"



namespace operations_research {
void BasicExample() {
  LOG(INFO) << "Google OR-Tools version : " << OrToolsVersion::VersionString();

  

  

  std::unique_ptr<MPSolver> solver(MPSolver::CreateSolver("GLOP"));
  if (!solver) {
    LOG(WARNING) << "Could not create solver GLOP";
    return;
  }
  


  

  

  MPVariable* const x = solver->MakeNumVar(0.0, 1, "x");
  MPVariable* const y = solver->MakeNumVar(0.0, 2, "y");

  LOG(INFO) << "Number of variables = " << solver->NumVariables();
  


  

  

  const double infinity = solver->infinity();
  MPConstraint* const ct = solver->MakeRowConstraint(-infinity, 2.0, "ct");
  ct->SetCoefficient(x, 1);
  ct->SetCoefficient(y, 1);

  LOG(INFO) << "Number of constraints = " << solver->NumConstraints();
  


  

  

  MPObjective* const objective = solver->MutableObjective();
  objective->SetCoefficient(x, 3);
  objective->SetCoefficient(y, 1);
  objective->SetMaximization();
  


  

  LOG(INFO) << "Solving with " << solver->SolverVersion();
  const MPSolver::ResultStatus result_status = solver->Solve();
  


  

  

  LOG(INFO) << "Status: " << result_status;
  if (result_status != MPSolver::OPTIMAL) {
    LOG(INFO) << "The problem does not have an optimal solution!";
    if (result_status == MPSolver::FEASIBLE) {
      LOG(INFO) << "A potentially suboptimal solution was found";
    } else {
      LOG(WARNING) << "The solver could not solve the problem.";
      return;
    }
  }

  LOG(INFO) << "Solution:";
  LOG(INFO) << "Objective value = " << objective->Value();
  LOG(INFO) << "x = " << x->solution_value();
  LOG(INFO) << "y = " << y->solution_value();
  


  

  LOG(INFO) << "Advanced usage:";
  LOG(INFO) << "Problem solved in " << solver->wall_time() << " milliseconds";
  LOG(INFO) << "Problem solved in " << solver->iterations() << " iterations";
  

}
}  


int main(int argc, char* argv[]) {
  InitGoogle(argv[0], &argc, &argv, true);
  absl::SetFlag(&FLAGS_stderrthreshold, 0);
  operations_research::BasicExample();
  return EXIT_SUCCESS;
}

File: test/ok/basic_example.ts
import { operations_research as op } from '../../src'
function test()
{
    console.log("Google OR-Tools version : " + op.OrToolsVersion.VersionString());

    let solver = op.MPSolver.CreateSolver("GLOP");
    if (!solver)
    {
        console.log("Could not create solver GLOP");
        return;
    }
    let x = solver.MakeNumVar(0.0, 1, "x");
    let y = solver.MakeNumVar(0.0, 2, "y");
    console.log("Number of variables = " + solver.NumVariables());
    let infinity = op.MPSolver.infinity();
    let ct = solver.MakeRowConstraint(-infinity, 2.0, "ct");
    ct.SetCoefficient(x, 1);
    ct.SetCoefficient(y, 1);
    console.log("Number of constraints = " + solver.NumConstraints());
    let objective = solver.MutableObjective();
    objective.SetCoefficient(x, 3);
    objective.SetCoefficient(y, 1);
    objective.SetMaximization();

    console.log("Solving with " + solver.SolverVersion());
    let result_status = solver.Solve();
    console.log("Status: " + result_status);
    if (result_status != op.MPSolver.ResultStatus.OPTIMAL)
    {
        console.log("The problem does not have an optimal solution!");
        if (result_status == op.MPSolver.ResultStatus.FEASIBLE)
        {
            console.log("A potentially suboptimal solution was found");
        }
        else
        {
            console.log("The solver could not solve the problem.");
            return;
        }
    }
    console.log("Solution:");
    console.log("Objective value = " + objective.Value());
    console.log("x = " + x.solution_value());
    console.log("y = " + y.solution_value());
    console.log("Advanced usage:");
    console.log("Problem solved in " + solver.wall_time() + " milliseconds");
    console.log("Problem solved in " + solver.iterations() + " iterations");
}
test();
